### PR TITLE
API Wrapper: filter expressions on catalog fetch

### PR DESCRIPTION
**Pull request by kamilczaja**
_What issues does this PR close?_
Closes internal issue.

```[tasklist]
### Checklist
- [x] The PR title is short and expressive.
- [x] I have updated the CHANGELOG.md. See [changelog_update.md](https://github.com/sovity/authority-portal/tree/main/docs/dev/changelog_updates.md) for more information.
- [ ] I have updated the Deployment Migration Notes Section in the CHANGELOG.md for any configuration / external API changes.
- [x] I have performed a **self-review**
```

### CODE REVIEW

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseResource.java [hunk 1]
```diff
@@ -43,4 +52,14 @@ public interface UseCaseResource {
...
     List<String> getSupportedFunctions();
+
+    @POST
+    @Path("pages/catalog-page/data-offers")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Operation(description = "Fetch a connector's data offers")
+    List<UiDataOffer> queryCatalog(
+        @Valid @NotNull
+        CatalogQueryParams catalogQueryParams
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseResource.java:63]: CatalogQuery

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,8 @@ please see [changelog_updates.md](docs/dev/changelog_updates.md).
 
 #### Minor Changes
 
+- API Wrapper: Added an use case endpoint to fetch the catalog with filter expressions
```
Comment (richardtreier) [inline CHANGELOG.md:18]: reword to "API Wrapper Use Case API: Catalog Endpoint"

File: docs/api/sovity-edc-api-wrapper.yaml [hunk 3]
```diff
@@ -436,6 +436,27 @@ paths:
                 type: array
                 items:
                   type: string
+  /wrapper/use-case-api/pages/catalog-page/data-offers:
```
Comment (richardtreier) [inline docs/api/sovity-edc-api-wrapper.yaml:439]: Don't forget to regenerate this

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseResource.java [hunk 4]
```diff
@@ -43,4 +52,14 @@ public interface UseCaseResource {
     @Produces(MediaType.APPLICATION_JSON)
     @Operation(description = "List available functions in policies, prohibitions and obligations.")
     List<String> getSupportedFunctions();
+
+    @POST
+    @Path("pages/catalog-page/data-offers")
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseResource.java:57]: "/catalog"

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpression.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@RequiredArgsConstructor
+@Schema(description = "Filter expression for catalog filtering", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpression.java:11]: Generic expression for filtering the data offers in the catalog

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpression.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+public class CatalogFilterExpression {
+    @Schema(description = "Left Operand", requiredMode = Schema.RequiredMode.REQUIRED)
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpression.java:13]: "Asset proprty name", add examples: ID, Title, custom property like "e.g. id, name, http://custom-property"?

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpressionLiteralType.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package de.sovity.edc.ext.wrapper.api.usecase.model;
+
+import io.swagger.v3.oas.annotations.media.Schema;
+
+@Schema(description = "Value type of an asset selector criterion right expression value", enumAsRef = true)
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogFilterExpressionLiteralType.java:5]: asset selector criterion?

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogQueryParams.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@RequiredArgsConstructor
+@Schema(description = "Catalog query parameters")
+public class CatalogQueryParams {
+    @Schema(description = "Target EDC DSP endpoint URL", requiredMode = Schema.RequiredMode.REQUIRED)
+    private String targetEdc;
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogQueryParams.java:14]: needs to have "connectorEndpoint" in it, e.g. targetEdcConnectorEndpoint or connectorEndpoint

File: extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogQueryParams.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    @Schema(description = "Limit the number of results", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
+    private Integer limit;
+
+    @Schema(description = "Offset for returned results, e.g. start at result 2", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
+    private Integer offset;
+
+    @Schema(description = "Filter expression for catalog filtering", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
+    private CatalogFilterExpression filterExpression;
```
Comment (richardtreier) [inline extensions/wrapper/wrapper-api/src/main/java/de/sovity/edc/ext/wrapper/api/usecase/model/CatalogQueryParams.java:23]: nur eine? Nicht mehrere?

File: extensions/wrapper/wrapper/src/main/java/de/sovity/edc/ext/wrapper/WrapperExtensionContextBuilder.java [hunk 10]
```diff
@@ -263,9 +272,11 @@ public static WrapperExtensionContext buildContext(
                 transferProcessStateService
         );
         var supportedPolicyApiService = new SupportedPolicyApiService(policyEngine);
+        var useCaseCatalogApiService = new UseCaseCatalogApiService(uiDataOfferBuilder, dspCatalogService, filterExpressionMapper);
         var useCaseResource = new UseCaseResourceImpl(
                 kpiApiService,
-                supportedPolicyApiService
+                supportedPolicyApiService,
+                useCaseCatalogApiService
         );
```
Comment (richardtreier) [inline extensions/wrapper/wrapper/src/main/java/de/sovity/edc/ext/wrapper/WrapperExtensionContextBuilder.java:280]: image

start newlining as these expressions are pretty long

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+    @Test
+    void shouldFetchFilteredDataOffersWithEq() {
+        // arrange
+        setupAssets();
+        buildContractDefinition(policyId, assetId1, "cd-1");
+        buildContractDefinition(policyId, assetId2, "cd-2");
+
+        // act
+        var catalogQueryParamsEq = buildCatalogQueryParams(assetId1, CatalogFilterExpressionOperator.EQ);
```
Comment (richardtreier) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java:54]:
```java
var filter = criterion("id", CatalogFilterExpressionOperator.EQ, assetId1)
var filter2 = criterion("id", CatalogFilterExpressionOperator.IN, List.of(assetId1))
```

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+    @Test
+    void shouldFetchFilteredDataOffersWithLike() {
+        // arrange
+        setupAssets();
+        buildContractDefinition(policyId, assetId1, "cd-1");
+        buildContractDefinition(policyId, assetId2, "cd-2");
+
+        // act
+        var catalogQueryParamsEq = buildCatalogQueryParams(assetId2, CatalogFilterExpressionOperator.LIKE);
```
Comment (richardtreier) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java:72]: like must be only part of the string

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+@ApiTest
+@ExtendWith(EdcExtension.class)
+public class UseCaseApiWrapperTest {
+    private EdcClient client;
+
+    private String assetId1;
+    private String assetId2;
+    private String policyId;
```
Comment (richardtreier) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java:37]: increased abstraction, what are the values?

File: extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        buildContractDefinition(policyId, assetId1, "cd-1");
+        buildContractDefinition(policyId, assetId2, "cd-2");
+
+        // act
+        var catalogQueryParamsEq = buildCatalogQueryParamsWithLimit(1, 0);
+        var dataOffers = client.useCaseApi().queryCatalog(catalogQueryParamsEq);
+
+        // assert
+        assertThat(dataOffers).hasSize(1);
+        assertThat(dataOffers.get(0).getAsset().getAssetId()).matches(assetId1 + "|" + assetId2);
```
Comment (richardtreier) [inline extensions/wrapper/wrapper/src/test/java/de/sovity/edc/ext/wrapper/api/usecase/UseCaseApiWrapperTest.java:111]:
```java
assertThat(dataOffers)
  .extracting(it -> it.getAsset().getAssetId())
  .containsExactlyInAnyOrder(assetId1, assetId2)
```

File: utils/catalog-parser/src/main/java/de/sovity/edc/utils/catalog/DspCatalogService.java [hunk 15]
```diff
@@ -34,17 +34,36 @@ public DspCatalog fetchDataOffers(String endpoint) throws DspCatalogServiceExcep
         return dspDataOfferBuilder.buildDataOffers(endpoint, catalogJson);
     }
 
+    public DspCatalog fetchDataOffersWithFilters(String endpoint, QuerySpec querySpec) {
```
Comment (richardtreier) [inline utils/catalog-parser/src/main/java/de/sovity/edc/utils/catalog/DspCatalogService.java:37]: change the old function to:

```kt
    public DspCatalog fetchDataOffers(String endpoint) throws DspCatalogServiceException {
        return fetchDataOffers(endpoint, QuerySpec.max());
    }
```

**Review by richardtreier — CHANGES_REQUESTED** [review]

**Review by ununhexium — APPROVED** [review]
LGTM.

**Pull request merged**